ostd/src/device/dispatcher/io_mem.rs

```diff
@@ -0,0 +1,234 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Io Memory access dispatcher.
+
+use alloc::{borrow::ToOwned, vec::Vec};
+use core::ops::Range;
+
+use log::{debug, info};
+use spin::Once;
+
+use crate::{
+    device::io_mem::IoMem,
+    mm::{CachePolicy, PageFlags},
+    util::vaddr_alloc::VirtAddrAllocator,
+};
+
+/// I/O memory dispatcher that allocates I/O access to devices via MMIO.
+pub struct IoMemDispatcher {
+    allocators: Once<Vec<VirtAddrAllocator>>,
+}
+
+impl IoMemDispatcher {
+    /// Get the IO memory access with `range`. If the range is not available, then the return value will be None.
+    pub fn get(&self, range: Range<usize>) -> Option<IoMem> {
+        let mut result = None;
+        for allocator in self.allocators.get().unwrap().iter() {
+            let allocator_range = allocator.fullrange();
+            if allocator_range.start >= range.end || allocator_range.end <= range.start {
+                continue;
+            }
+
+            let start_addr = range.start;
+            let end_addr = range.end;
+            debug!("Allocating MMIO range:{:x?}..{:x?}", start_addr, end_addr);
+            if allocator.alloc_specific(range).is_err() {
+                return None;
+            }
+
+            // SAFETY: The created IoMem is guaranteed not to access physical memory or system device I/O
+            unsafe {
+                result = Some(IoMem::new(
+                    start_addr..end_addr,
+                    PageFlags::RW,
+                    CachePolicy::Uncacheable,
+                ))
+            }
+            break;
+        }
+        result
+    }
+
+    /// Remove access to a specific memory IO range.  
+    ///
+    /// All drivers in OSTD must use this method to prevent peripheral drivers from accessing illegal memory IO range.
+    pub(crate) fn remove(&self, range: Range<usize>) {
+        for allocator in self.allocators.get().unwrap().iter() {
+            let allocator_range = allocator.fullrange();
+            if allocator_range.start >= range.end || allocator_range.end <= range.start {
+                continue;
+            }
+
+            let start_addr = range.start;
+            let end_addr = range.end;
+            debug!("Removing MMIO range:{:x}..{:x}", start_addr, end_addr);
+
+            // Remove I/O memory.
+            let _ = allocator.alloc_specific(range);
+            return;
+        }
+    }
+
+    /// Recycle an MMIO range.
+    #[allow(dead_code)]
+    pub(crate) fn recycle(&self, range: Range<usize>) {
+        for allocator in self.allocators.get().unwrap().iter() {
+            let allocator_range = allocator.fullrange();
+            if allocator_range.start >= range.end || allocator_range.end <= range.start {
```

Comment:
The process of finding the allocator can be refactored to avoid duplication of code in `get`/`remove`/`recycle`.